### ISSUE TITLE
Systematic solver failure for 3D case on latest master

### ISSUE THREAD

**Issue by tridelat** [erdc/proteus#626]
The latest master causes failures once per time step on the twp model (RANS2P) when running a 3D case, and this considerably increase the computational cost of running a simulation. It happens for a simple air/water 3D cuboidal domain with a still water level (free slip walls and atmophere BC on top).
I use the asm solver and run the case as follows:
`parun [case]_so.py -O [path to pets.options.asm`

It runs fine up to 5d5f1f5bb6e7934af33e7551749f11dd9e709186.
The next 2 commits, a052d8755c80549b7c0a3600cdfc52e7212706ee and 767d4e83ec6f1c34191ba08fdeb67fa0372ea701, the case does not run due to an error in `calculateJacobian()` of RANS2P (189 positional arguments given instead of 190).
From commit d99cf7539a717f67b9db446a938424261622fe33, it runs but gives me the failures at each time step. 

@cekees @alistairbntl @manuel-quezada , any idea what could cause this issue?

**Comment by alistairbntl** [comment]
@tridelat  - I've been working on the Marin problem the past few days and I've noticed a few issues too.  Perhaps most important, I see that the Newton iteration is not converging quadratically.  I suspect this is related to your issue and likely means the Jacobian and residual are inconsistent.
What is your setting for the NONCONSERVATIVE_FORM in the RANS2P Coefficients?

**Comment by manuel-quezada** [comment]
I agree with @alistairbntl that this sounds like an inconsistency between the residual and the Jacobian on RANS2P. Do you have any 2D example in some repository or branch we can run?

**Comment by manuel-quezada** [comment]
These models: RANS2P and RANS3PF have the possibility of considering different terms to model different physics. In other words, we have the chance to add terms to the equations if we consider stuff like turbulence, Darcy's forces, particles (in the case of RANS3PF) and other things. In addition, we can consider different types of boundary conditions. All these "extra" terms are on top of the standard terms for the NS equations. First, we need to try to identify which of these terms is the one that is creating the problem (or the inconsistency between the residual and the Jacobian). I suggest we try to turn off everything that is not part of the standard form of the NS equations. If that works then we can start putting back these terms one by one.

**Comment by cekees** [comment]
@tridelat could you commit the case to a repository so we can try to reproduce the failure? I've tried the marin 3D test in conservative and nonconservative form, in parallel and serial, and with superlu and with asm. The Newton iteration is converging quadratically in all those cases. Maybe @alistairbntl and @manuel-quezada could try to reproduce the failure. Are you running on copper?

**Comment by tridelat** [comment]
@alistairbntl the NON_CONSERVATIVE_FORM is set to the default (0) for RANS2P in the test case I ran.
@manuel-quezada @cekees I pushed a version that creates the failures here: https://github.com/erdc/air-water-vv/tree/bug_asm/3d/bug_asm
It fails only once per time step (after the first few time steps), and the simulation as a whole does not fail but it makes it very slow. It seems to be okay with superlu, but not with asm. I ran it in parallel with 2 processors on my laptop (Arch Linux) and hydra (Centos) and got the same issue. I could also see the issue on copper when trying to run a much larger test case (with floating body + moorings), where the simulation finished normally but was very slow due to the systematic failures.

**Comment by cekees** [comment]
I was able to reproduce the problem on my linux machine and confirm that it runs with superlu in serial and superlu_dist in parallel, which would suggest the issue is in some of the recent linear algebra upgrades. The nonlinear residuals aren't decreasing as fast as I'd like even with a direct factorization. The first iteration shows the usual quadratic decrease but sometimes the second iteration doesn't produce a large decrease.

**Comment by alistairbntl** [comment]
@cekees: It sounds like my errors on the `bug_asm` problem are different from yours.   (I'm on Proteus `master`: 867efbd, `hashstack`: fbf0d9 and air-water-vv `bug_asm`: cbb79ea).  Please let me know if you're experiencing something similar to the following:

When I run `bug_asm` using superlu_dist (two processors), I frequently get a residual increase during the first Navier-Stokes non-linear solve.  Because the solver is direct, the linear solve only takes a single step .  
At times (such as model substep t=6.51708e-01) the non-linear solver's behavior becomes more erratic (e.g. multiple residual increases and a large number of iterations before convergence.)  There is also two points in the simulation at which the non-linear solver fails altogether. 
Towards the end of the simulation, the residual increase become less frequent, but they still pop up from time to time. 

When I run `bug_asm` with asm (two processors), the Navier Stokes I see similar things happen, but more frequently and dramatically.  For instance, the non-linear solver fails 10 times.   The linear solver also seems have trouble and 100+ gmres iterations are frequent.

Ultimately both simulations run to completion, and seem to generate the same solution profile in paraview.  The final solution looks as follows:

![solution_ex](https://user-images.githubusercontent.com/10620679/31736130-e163a476-b411-11e7-8dff-ccd579bf134d.png)

**Comment by tridelat** [comment]
I am confirming that the asm solver acts as intended on air-water-vv cbb79ea after the #628 merge, so this issue can be closed
